src/pages/VaultsPage/VaultTransactions.tsx

```diff
@@ -1,62 +1,285 @@
+  }: VaultAndTransactions
+): Promise<RawTransaction[]> {
+  if (!(token_address || simple_token_address)) return [];
+  const token = token_address || (simple_token_address as string);
+  const withdrawFilter = contracts.router.filters.WithdrawFromVault();
+  const allWithdrawEvents = await contracts.router.queryFilter(
+    withdrawFilter,
+    deployment_block
+  );
+  const eventTxInfo = await Promise.all(
+    allWithdrawEvents.map(e => e.getTransactionReceipt())
+  );
+  const erc20Contract = contracts.getERC20(token);
+  const withdrawEvents = allWithdrawEvents.filter(
+    (_, idx) => eventTxInfo[idx].to.toLowerCase() === vault_address
+  );
+
+  const withdraws: RawTransaction[] = [];
+  for (const event of withdrawEvents) {
+    const tokenFilter = erc20Contract.filters.Transfer(event.args.vault);
+    const tokenTransferEvents = await erc20Contract.queryFilter(
+      tokenFilter,
+      event.blockNumber,
+      event.blockNumber
+    );
+    const transferEvent = tokenTransferEvents.find(
+      e => e.transactionHash === event.transactionHash
+    );
+    if (!transferEvent) continue;
+    const block = await event.getBlock();
+    const txDetails = vault_transactions.find(
+      tx => tx.tx_hash === event.transactionHash
+    );
+    if (
+      txDetails?.tx_type === 'Withdraw' &&
+      txDetails.profile?.address === transferEvent.args.to.toLowerCase()
+    ) {
+      const user = txDetails.profile.users.pop();
+      withdraws.push({
+        block: event.blockNumber,
+        type: 'Withdraw',
+        details: `By ${user?.name || transferEvent.args.from}`,
+        amount: transferEvent.args.value.div(BigNumber.from(10).pow(decimals)),
+        date: DateTime.fromSeconds(block.timestamp).toFormat('DD'),
+        hash: event.transactionHash,
+      });
+      continue;
+    }
+    /* log from outside app
+    withdraws.push({
+      block: event.blockNumber,
+      type: 'Withdraw',
+      details: `By ${transferEvent.args.from}`,
+      amount: event.args.amount.div(BigNumber.from(10).pow(decimals)),
+      date: DateTime.fromSeconds(block.timestamp).toFormat('DD'),
+      hash: event.transactionHash,
+    });
+    */
+  }
+  return withdraws;
+}
+
+interface RawDistributionTx extends RawTransaction {
+  circleId: number;
+  circle: string;
+}
+async function getDistributionEvents(
+  contracts: ContractsReadonly,
+  { vault_address, deployment_block, vault_transactions }: VaultAndTransactions
+): Promise<RawDistributionTx[]> {
+  const distroFilter = contracts.distributor.filters.EpochFunded(vault_address);
+  const distroEvents = await contracts.distributor.queryFilter(
+    distroFilter,
+    deployment_block
+  );
+
+  const distributions: RawDistributionTx[] = [];
+  for (const event of distroEvents) {
+    const block = await event.getBlock();
+    const txDetails = vault_transactions.find(
+      tx => tx.tx_hash === event.transactionHash
+    );
+    const { name: circleName, id: circleId } = await getCircle(
```

Comment:
you could avoid an extra request here by adding these two fields to the VaultAndTransactions query. yes, then you're depending on the vault_transactions data rather than decoding the circle id from the event, but these should be the same anyway